File: leetcode-questions/1079-letter-tile-possibilities/1079-letter-tile-possibilities.py
class Solution:
    def __init__(self):
        self.tiles = ""
        self.seen = set()

    def backtrack(self, curr_tiles, index_set):
        if curr_tiles in self.seen:
            return 0

        self.seen.add(curr_tiles)
        sequences = 1
        for index, letter in enumerate(self.tiles):
            curr_tiles += letter
            if curr_tiles not in self.seen and index not in index_set:
                index_set.add(index)
                sequences += self.backtrack(curr_tiles, index_set)
                index_set.remove(index)
            curr_tiles = curr_tiles[:-1]

        return sequences

    def numTilePossibilities(self, tiles: str) -> int:
        self.tiles = tiles
        all_sequences = 0
        for index, letter in enumerate(self.tiles):
            all_sequences += self.backtrack(letter, set([index]))

        return all_sequences

File: js(ts)-solutions/1079LetterTilePossibilities.ts
function numTilePossibilities(tiles: string): number {
    let seen: Set<string> = new Set<string>();
    let N: number = tiles.length;

    const countSequences = (currTiles: string, indexSet: Set<number>) => {
        if (seen.has(currTiles)) return

        if (currTiles.trim() !== '') seen.add(currTiles);
        for (let i = 0; i < N; i++) {
            currTiles += tiles[i];
            if (!seen.has(currTiles) && !indexSet.has(i)) {
                indexSet.add(i);
                countSequences(currTiles, indexSet);
                indexSet.delete(i);
            }
            currTiles = currTiles.slice(0, -1);
        }
    };
    countSequences('', new Set<number>());
    return seen.size;
};